pybamm/solvers/solution.py

```diff
@@ -11,132 +10,145 @@
 
     def __init__(
-        self, t, y, t_event=None, y_event=None, termination="final time", copy_this=None
+        self,
+        all_ts,
+        all_ys,
+        model,
+        all_inputs,
+        t_event=None,
+        y_event=None,
+        termination="final time",
     ):
-        self.t = t
-        self.y = y
+        if not isinstance(all_ts, list):
+            all_ts = [all_ts]
+        if not isinstance(all_ys, list):
+            all_ys = [all_ys]
+        self.all_ts = all_ts
+        self.all_ys = all_ys
+
         self._t_event = t_event
         self._y_event = y_event
         self._termination = termination
-        if copy_this is None:
-            # initialize empty inputs and model, to be populated later
-            self._inputs = pybamm.FuzzyDict()
-            self.model = pybamm.BaseModel()
-            self.set_up_time = None
-            self.solve_time = None
-            self.integration_time = None
-            self.has_symbolic_inputs = False
+
+        # Set up inputs
+        if not isinstance(all_inputs, list):
+            for key, value in all_inputs.items():
+                if isinstance(value, numbers.Number):
+                    all_inputs[key] = np.array([value])
+            all_inputs = [all_inputs]
+        self.all_inputs = all_inputs
+        self.has_symbolic_inputs = any(
+            isinstance(v, casadi.MX) for v in all_inputs[0].values()
+        )
+
+        # Set up model
+        self._model = model
+
+        # Copy the timescale_eval and lengthscale_evals if they exist
+        if hasattr(model, "timescale_eval"):
+            self.timescale_eval = model.timescale_eval
         else:
-            self._inputs = copy.copy(copy_this.inputs)
-            self.model = copy_this.model
-            self.set_up_time = copy_this.set_up_time
-            self.solve_time = copy_this.solve_time
-            self.integration_time = copy_this.integration_time
-            self.has_symbolic_inputs = copy_this.has_symbolic_inputs
+            self.timescale_eval = model.timescale.evaluate()
+        # self.timescale_eval = model.timescale_eval
+        if hasattr(model, "length_scales_eval"):
+            self.length_scales_eval = model.length_scales_eval
+        else:
+            self.length_scales_eval = {
+                domain: scale.evaluate()
+                for domain, scale in model.length_scales.items()
+            }
+
+        self.set_up_time = None
+        self.solve_time = None
+        self.integration_time = None
 
         # initiaize empty variables and data
         self._variables = pybamm.FuzzyDict()
         self.data = pybamm.FuzzyDict()
 
+        # Add self as sub-solution for compatibility with ProcessedVariable
+        self._sub_solutions = [self]
+
+        # Solution now uses CasADi
+        pybamm.citations.register("Andersson2019")
+
     @property
     def t(self):
         "Times at which the solution is evaluated"
-        return self._t
+        try:
```

Comment:
what happens if I call `solution.t` (or `solution.y`) then take a solver step and call `solution.t` again? does it get updated to include the new times/states or does it miss them since it already has the attribute `_t`?

Reply:
I've actually removed `append` and made it so `__add__` returns a new object (i.e. the two solutions being added are unchanged), which I think is more intuitive and not slower

Reply:
ah yeah I see that now, thanks